Add TORCH_CHECK_NOT_IMPLEMENTED/c10::NotImplementedError; make dispatch use it

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53617 Marginally improve pytest collection for top-level test files
* #53611 Fix inaccurate dispatch table for fill_
* #53610 Convert a few more checks to raise NotImplementedError
* #53397 Add Meta support for empty_strided
* #53396 Add debug only layout assert for empty_cpu
* **#53377 Add TORCH_CHECK_NOT_IMPLEMENTED/c10::NotImplementedError; make dispatch use it**

My underlying goal is I want to make the test suite ignore
NotImplementedError without failing when bringing up a backend (meta)
that doesn't have very many functions implemented.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26850766](https://our.internmc.facebook.com/intern/diff/D26850766)

ezyang committed:
Add TORCH_CHECK_NOT_IMPLEMENTED/c10::NotImplementedError; make dispatch use it

My underlying goal is I want to make the test suite ignore
NotImplementedError without failing when bringing up a backend (meta)
that doesn't have very many functions implemented.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

personal note: we should update the autograd backward formula that are not implemented to throw this as well.

ezyang committed:
Update on "Add TORCH_CHECK_NOT_IMPLEMENTED/c10::NotImplementedError; make dispatch use it"


My underlying goal is I want to make the test suite ignore
NotImplementedError without failing when bringing up a backend (meta)
that doesn't have very many functions implemented.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26850766](https://our.internmc.facebook.com/intern/diff/D26850766)

[ghstack-poisoned]
